rules/dynamic_archive.bzl

```diff
@@ -0,0 +1,215 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Dynamic archive macro loads repos from archives with dynamic tarballs but
+stable contents for OpenTitan.
+"""
+
+load(
+    "@bazel_tools//tools/build_defs/repo:utils.bzl",
+    "patch",
+    "read_netrc",
+    "read_user_netrc",
+    "update_attrs",
+    "use_netrc",
+    "workspace_and_buildfile",
+)
+
+def _get_auth(ctx, urls):
+    """Given the list of URLs obtain the correct auth dict."""
+    if ctx.attr.netrc:
+        netrc = read_netrc(ctx, ctx.attr.netrc)
+    else:
+        netrc = read_user_netrc(ctx)
+    return use_netrc(netrc, urls, ctx.attr.auth_patterns)
+
+def _check_dir_sha(ctx):
+    """Utility function checks the sha256 of a directory instead of an archive
+
+    It assumes the parameters `sha256` and `output` to be present in
+    `ctx.attr`; the latter possibly with value None.
+
+    Args:
+      ctx: The repository context of the repository rule calling this utility
+      function.
+
+    Returns the sha256 of a gzipped tarball of the directory contents.
+    """
+    command = "find . -type f -exec sha256sum {} \\; | sort -k 2 | sha256sum"
```

Comment:
`find ... -exec` is almost never what you want, as it is slow (it execs a new `sha256sum` for each file).  Instead, you want to pipe `find` into `xargs` and exec the target process once:
```
find . -type f -print0 | xargs -0 sha256sum | sort -k 2 | sha256sum
```

Note: `-print0` (and the corresponding `-0` on `xargs`) delimits the output with `\0` chars instead of spaces thus ppreventing any oddities when encountering filenames that have spaces in them.

For reference, here's how the the two different shell pipelines compare on my system (against boringssl):
```
cfrantz@cfrantz-glapstation:~/src/tmp$ time find . -type f -exec sha256sum {} \; | sort -k 2 | sha256sum
9b0e6195cf7270149f7719467c284f4882bdfb7a6c288fb9ca6618ef2e3dc83f  -

real	0m14.130s
user	0m8.591s
sys	0m6.099s
cfrantz@cfrantz-glapstation:~/src/tmp$ time find . -type f -print0 | xargs -0 sha256sum | sort -k 2 | sha256sum
9b0e6195cf7270149f7719467c284f4882bdfb7a6c288fb9ca6618ef2e3dc83f  -

real	0m0.822s
user	0m0.706s
sys	0m0.205s
```